mautic/mautic
# Adds SparkPost support

[//]: # (Please answer the following questions. )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | Y
| Related user documentation PR URL | todo
| Related developer documentation PR URL | todo
| Issues addressed (#s or URLs) | Fixes #1488 and fixes #1088
| BC breaks? | n
| Deprecations? | n

Partial credit goes to @jfgrissom for getting this started!

[//]: # (**Note that all new features should have a related user and/or developer documentation PR in their respective repositories.**)

[//]: # (### Required)
#### Description:

This PR adds support for Sparkpost's API - it leverages Sparkpost's substitution and metadata feature to support tokenized emails. 

It also has webhook support for bounce, spam_complaint, out_of_band, policy_rejection, list_unsubscribe, and link_unsubscribe. (Register the webhook to http://your-mautic.com/mailer/sparkpost/callback)

This PR supersedes #1715.

It also fixes a bug with HTTP mailers (mandrill and now this) that did not register bounces/unsubscribes due to the API changes for registering a DNC status.

#### Steps to test this PR:
1. Apply the PR and clear cache
2. Go to Configuration -> Email Settings and choose Sparkpost along with enter the API key from Sparkpost admin
3. Create a new segment  with a handful of contacts in it
4. Create a new segment email that includes tokens (contactfield tokens for example)
5. Send the email - Sparkpost should deliver them with the tokens replaced appropriately
6. To test the callback - configure a webhook in sparkpost with http://your-mautic.com/mailer/sparkpost/callback (must be world accessible)
7. Send a bogus email through Mautic
8. It should register the lead as bounced/unsubscribed

### As applicable
1. To test before the PR requires a Mandrill account - after the PR the bug is fixed so not applicable. As long as the PR is a success - the bug is fixed ;-)
